crypto/evp/p_lib.c

```diff
@@ -2495,3 +2495,119 @@ int EVP_PKEY_get_field_type(const EVP_PKEY *pkey)
     return 0;
 }
 #endif
+
+/*- All methods below can also be used in FIPS_MODULE */
+
+int EVP_SKEY_fromdata(EVP_SKEY *pskey, const OSSL_PARAM *params)
+{
+    void *keydata = NULL;
+
+    if (pskey == NULL || pskey->keydata != NULL)
+        return -1;
+
+    if ((keydata = evp_keymgmt_newdata(pskey->keymgmt)) == NULL
+        || !evp_keymgmt_import(pskey->keymgmt, keydata, OSSL_KEYMGMT_SELECT_ALL, params)) {
```

Comment:
I would consider hainv evp_keymgmt_import directly return keydata instead of having to call evp_keymgmt_newdata() first.

At least for the pkcs11-provider this pattern makes things unnecessarily hard for asymmetric keys as we have to create a fake empty key and then handle cases where an application may try to use it, and we also need to handle the case someone try to import data on an existing key which is not a well defined operation (should not be allowed I think). In short it needlessly multiplies error checking for us, and it is unclear what advantage there is in splitting container creation and actual import in two calls.

Reply:
I think that your proposal doesn't match the current EVP_KEYMGMT implementation though it may make sense for PKCS#11